packages/runtime/container-runtime/src/summarizer.ts

```diff
@@ -238,10 +257,54 @@ export class Summarizer extends EventEmitter implements ISummarizer {
     }
 
     public readonly summarizeOnDemand: ISummarizer["summarizeOnDemand"] = (...args) => {
-        if (this._disposed || this.runningSummarizer === undefined || this.runningSummarizer.disposed) {
-            throw Error("Summarizer is not running or already disposed.");
+        try {
+            if (this._disposed) {
+                throw Error("Summarizer is already disposed.");
+            }
+            if (this.runtime.summarizerClientId !== undefined &&
+                this.runtime.summarizerClientId !== this.runtime.clientId) {
+                throw Error("On-demand summary attempted while an elected summarizer is present");
+            }
+            const builder = new OnDemandSummarizeResultBuilder();
+            if (this.runningSummarizer && !this.runningSummarizer.disposed) {
+                // Summarizer is already running. Go ahead and start.
+                builder.summarizerStarted.resolve({ success: true, data: undefined });
+                return this.runningSummarizer.summarizeOnDemand(builder, ...args);
+            }
+
+            // Summarizer isn't running, so we need to start it, which is an async operation.
+            // Manage the promise related to creating the cancellation token here.
+            // The promises related to starting, summarizing,
+            // and submitting are communicated to the caller through the results builder.
+            this.runningSummarizer = undefined;
+            const coordinatorCreateP = this.runCoordinatorCreateFn(this.runtime);
+
+            coordinatorCreateP.then((runCoordinator) => {
```

Comment:
So it's possible for caller to call summarizeOnDemand() twice in a row and both instances reach here, and thus call this.start(). Is it Ok? Should we properly resolve this race (by ensuring that second call will fail for first one to resolve)? Or fail the call as wrong usage pattern?

Reply:
Good point. Looks like it's also possible to get into this situation by calling run() repeatedly. I'll add a flag that will cause start() to fail if it starts again without having completed. I can't think of a reason why this needs to be a promise that can be awaited.

Reply:
sure, as long as there is some protection, I'm fine with it